File: cpp/lib/concept/type/role_type.cpp

/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "typedb/concept/type/role_type.hpp"
#include "typedb/common/exception.hpp"
#include "typedb/connection/transaction.hpp"

#include "../../common/macros.hpp"
#include "../../common/native.hpp"
#include "../../common/utils.hpp"
#include "../concept_factory.hpp"

namespace TypeDB {

RoleType::RoleType(_native::Concept* conceptNative)
    : Type(ConceptType::ROLE_TYPE, conceptNative) {}

std::string RoleType::getName() {
    CHECK_NATIVE(conceptNative);
    return Utils::stringFromNative(_native::role_type_get_name(conceptNative.get()));
}

std::string RoleType::getScope() {
    CHECK_NATIVE(conceptNative);
    return Utils::stringFromNative(_native::role_type_get_scope(conceptNative.get()));
}

std::string RoleType::getLabel() {
    return getScope() + ":" + getName();
}

VoidFuture RoleType::setLabel(Transaction& transaction, const std::string& newLabel) {
    CONCEPTAPI_CALL(VoidFuture, _native::role_type_set_label(ConceptFactory::getNative(transaction), conceptNative.get(), newLabel.c_str()));
}

bool RoleType::isAbstract() {
    CHECK_NATIVE(conceptNative);
    return _native::role_type_is_abstract(conceptNative.get());
}

VoidFuture RoleType::drop(Transaction& transaction) {
    CONCEPTAPI_CALL(VoidFuture, _native::role_type_delete(ConceptFactory::getNative(transaction), conceptNative.get()));
}

BoolFuture RoleType::isDeleted(Transaction& transaction) {
    CONCEPTAPI_CALL(BoolFuture, _native::role_type_is_deleted(ConceptFactory::getNative(transaction), conceptNative.get()));
}

// "Inherited"
ConceptPtrFuture<RoleType> RoleType::getSupertype(Transaction& transaction) {
    return ConceptPtrFuture<RoleType>(getSuperTypeFutureNative(transaction));
}

ConceptIterable<RoleType> RoleType::getSupertypes(Transaction& transaction) {
    return ConceptIterable<RoleType>(getSupertypesIteratorNative(transaction));
}

ConceptIterable<RoleType> RoleType::getSubtypes(Transaction& transaction, Transitivity transitivity) {
    return ConceptIterable<RoleType>(getSubtypesIteratorNative(transaction, transitivity));
}

ConceptPtrFuture<RelationType> RoleType::getRelationType(Transaction& transaction){
    CONCEPTAPI_FUTURE(RelationType, _native::role_type_get_relation_type(ConceptFactory::getNative(transaction), conceptNative.get()))}

ConceptIterable<RelationType> RoleType::getRelationTypes(Transaction& transaction) {
    CONCEPTAPI_ITER(RelationType, _native::role_type_get_relation_types(ConceptFactory::getNative(transaction), conceptNative.get()));
}

ConceptIterable<ThingType> RoleType::getPlayerTypes(Transaction& transaction, Transitivity transitivity) {
    CONCEPTAPI_ITER(ThingType, _native::role_type_get_player_types(ConceptFactory::getNative(transaction), conceptNative.get(), (_native::Transitivity)transitivity));
}

ConceptIterable<Relation> RoleType::getRelationInstances(Transaction& transaction, Transitivity transitivity) {
    CONCEPTAPI_ITER(Relation, _native::role_type_get_relation_instances(ConceptFactory::getNative(transaction), conceptNative.get(), (_native::Transitivity)transitivity));
}

ConceptIterable<Thing> RoleType::getPlayerInstances(Transaction& transaction, Transitivity transitivity) {
    CONCEPTAPI_ITER(Thing, _native::role_type_get_player_instances(ConceptFactory::getNative(transaction), conceptNative.get(), (_native::Transitivity)transitivity));
}

}  // namespace TypeDB

File: csharp/Concept/Type/RoleType.cs
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
 
using System.Collections.Generic;
using System.Linq;

using TypeDB.Driver;
using TypeDB.Driver.Api;
using TypeDB.Driver.Common;
using TypeDB.Driver.Concept;
using static TypeDB.Driver.Api.IConcept.Transitivity;
using static TypeDB.Driver.Concept.Thing;
using static TypeDB.Driver.Concept.ThingType;

namespace TypeDB.Driver.Concept
{
    public class RoleType : Type, IRoleType
    {
        private Label? _label;

        public RoleType(Pinvoke.Concept nativeConcept)
            : base(nativeConcept)
        {
        }

        public override bool IsRoot()
        {
            return Pinvoke.typedb_driver.role_type_is_root(NativeObject);
        }

        public override bool IsAbstract()
        {
            return Pinvoke.typedb_driver.role_type_is_abstract(NativeObject);
        }

        public override Label Label
        {
            get
            {
                if (!_label.HasValue)
                {
                    _label = new Label(
                        Pinvoke.typedb_driver.role_type_get_scope(NativeObject),
                        Pinvoke.typedb_driver.role_type_get_name(NativeObject));
                }

                return _label.Value;
            }
        }

        public override VoidPromise Delete(ITypeDBTransaction transaction)
        {
            return new VoidPromise(Pinvoke.typedb_driver.role_type_delete(
                NativeTransaction(transaction), NativeObject).Resolve);
        }

        public override Promise<bool> IsDeleted(ITypeDBTransaction transaction)
        {
            return new Promise<bool>(Pinvoke.typedb_driver.role_type_is_deleted(
                NativeTransaction(transaction), NativeObject).Resolve);
        }

        public override VoidPromise SetLabel(ITypeDBTransaction transaction, string label)
        {
            return new VoidPromise(Pinvoke.typedb_driver.role_type_set_label(
                NativeTransaction(transaction), NativeObject, label).Resolve);
        }

        public override Promise<IType> GetSupertype(ITypeDBTransaction transaction)
        {
            return Promise<IType>.Map<Pinvoke.Concept, IType>(
                Pinvoke.typedb_driver.role_type_get_supertype(
                    NativeTransaction(transaction), NativeObject).Resolve,
                obj => new RoleType(obj));
        }

        public override IEnumerable<IType> GetSupertypes(ITypeDBTransaction transaction)
        {
            try
            {
                return new NativeEnumerable<Pinvoke.Concept>(
                    Pinvoke.typedb_driver.role_type_get_supertypes(
                        NativeTransaction(transaction), NativeObject))
                    .Select(obj => new RoleType(obj));
            } 
            catch (Pinvoke.Error e) 
            {
                throw new TypeDBDriverException(e);
            }
        }

        public override IEnumerable<IType> GetSubtypes(ITypeDBTransaction transaction)
        {
            return GetSubtypes(transaction, Transitive);
        }

        public override IEnumerable<IType> GetSubtypes(
            ITypeDBTransaction transaction, IConcept.Transitivity transitivity) 
        {
            try 
            {
                return new NativeEnumerable<Pinvoke.Concept>(
                    Pinvoke.typedb_driver.role_type_get_subtypes(
                        NativeTransaction(transaction), NativeObject, (Pinvoke.Transitivity)transitivity))
                    .Select(obj => new RoleType(obj));
            } 
            catch (Pinvoke.Error e) 
            {
                throw new TypeDBDriverException(e);
            }
        }

        public Promise<IRelationType> GetRelationType(ITypeDBTransaction transaction)
        {
            return Promise<IRelationType>.Map<Pinvoke.Concept, IRelationType>(
                Pinvoke.typedb_driver.role_type_get_relation_type(
                    NativeTransaction(transaction), NativeObject).Resolve,
                obj => new RelationType(obj));
        }

        public IEnumerable<IRelationType> GetRelationTypes(ITypeDBTransaction transaction)
        {
            try 
            {
                return new NativeEnumerable<Pinvoke.Concept>(
                    Pinvoke.typedb_driver.role_type_get_relation_types(
                        NativeTransaction(transaction), NativeObject))
                    .Select(obj => new RelationType(obj));
            } 
            catch (Pinvoke.Error e) 
            {
                throw new TypeDBDriverException(e);
            }
        }

        public IEnumerable<IThingType> GetPlayerTypes(ITypeDBTransaction transaction)
        {
            return GetPlayerTypes(transaction, Transitive);
        }

        public IEnumerable<IThingType> GetPlayerTypes(
            ITypeDBTransaction transaction, IConcept.Transitivity transitivity)
        {
            try
            {
                return new NativeEnumerable<Pinvoke.Concept>(
                    Pinvoke.typedb_driver.role_type_get_player_types(
                        NativeTransaction(transaction), NativeObject, (Pinvoke.Transitivity)transitivity))
                    .Select(obj => ThingTypeOf(obj));
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public IEnumerable<IRelation> GetRelationInstances(ITypeDBTransaction transaction)
        {
            return GetRelationInstances(transaction, Transitive);
        }

        public IEnumerable<IRelation> GetRelationInstances(
            ITypeDBTransaction transaction, IConcept.Transitivity transitivity)
        {
            try
            {
                return new NativeEnumerable<Pinvoke.Concept>(
                    Pinvoke.typedb_driver.role_type_get_relation_instances(
                        NativeTransaction(transaction), NativeObject, (Pinvoke.Transitivity)transitivity))
                    .Select(obj => new Relation(obj));
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public IEnumerable<IThing> GetPlayerInstances(ITypeDBTransaction transaction)
        {
            return GetPlayerInstances(transaction, Transitive);
        }

        public IEnumerable<IThing> GetPlayerInstances(
            ITypeDBTransaction transaction, IConcept.Transitivity transitivity)
        {
            try
            {
                return new NativeEnumerable<Pinvoke.Concept>(
                    Pinvoke.typedb_driver.role_type_get_player_instances(
                        NativeTransaction(transaction), NativeObject, (Pinvoke.Transitivity)transitivity))
                    .Select(obj => ThingOf(obj));
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }
    }
}
